### PR TITLE
Feature : allow users to link runs

### PR DESCRIPTION
**Pull request by WilliamHPNielsen**
This PR supersedes #1615 

I need to 

- [x] Take care of merge conflict resolving fallout :P
- [ ] ~Make the version 7 -> 8 upgrader populate the `parent_datasets` column with the default value `"[]"`~ No reason for this, since we've agreed to treat that column as optional
- [x] Update the "generate-fixture-dbs" repo with some version 7 DBs
- [x] Use those DBs in CI for this branch

and then we should be ready to merge.

@QCoDeS/core

### CODE REVIEW

File: qcodes/dataset/data_set.py [hunk 1]
```diff
@@ -497,6 +515,29 @@ def add_parameter(self, spec: ParamSpec):
                                   'Please use DataSet.set_interdependencies '
                                   'instead.')
 
+    def set_parent_dataset_links(self, links: List[Link]) -> None:
```
Comment (jenshnielsen) [inline qcodes/dataset/data_set.py]: Should this be a property setter?

Comment (WilliamHPNielsen) [reply qcodes/dataset/data_set.py]: Yes, that's probably more natural. The same holds true for `set_interdependencies`, but we can change that later.

File: qcodes/dataset/linked_datasets/links.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    @staticmethod
+    def validate_node(node_guid: str, node: str) -> None:
+        """
+        Validate that the guid given is a valid guid.
+
+        Args:
+            node_guid: the guid
+            node: either "head" or "tail"
+        """
+        if _guid_pattern.match(node_guid):
```
Comment (jenshnielsen) [inline qcodes/dataset/linked_datasets/links.py]: Should we not refactor this out into function in the guids.py (is_valid_guid or some such thing)

File: qcodes/dataset/measurements.py [hunk 3]
```diff
@@ -721,6 +732,27 @@ def _paramspecbase_from_strings(
 
         return tuple(depends_on), tuple(inf_from)
 
+    def register_parent(
+            self: T, parent: DataSet, link_type: str,
```
Comment (jenshnielsen) [inline qcodes/dataset/measurements.py]: So currently `link_type` can be anything?

File: qcodes/dataset/sqlite/queries.py [hunk 4]
```diff
@@ -1512,6 +1531,37 @@ def get_run_description(conn: ConnectionPlus, run_id: int) -> str:
...
+    """
+
+    # The RuntimeError comes from the atomic contextmanager as it rolls back
+    # the transaction. We assume that the rollback is caused by the column
+    # `parent_datasets` not being present. We find it fair to interpret this
+    # situation as "no parent datasets were registered"
+    #
+    # We cannot in general trust that NULLs will not appear in the column,
+    # so we must explicitly handle them.
+
```
Comment (jenshnielsen) [inline qcodes/dataset/sqlite/queries.py]: I would prefer to use pragma table info to check for the existence. How is a null value in the db handled here?

File: qcodes/dataset/measurements.py (hunk 3 above)
Comment (WilliamHPNielsen) [reply qcodes/dataset/measurements.py]: Yes. It's completely the responsibility of the users to ensure that they always write `"analysis"` and never `"analsyis"`, or at least that they can handle that ambiguity. Do you think we can do better than this?

File: qcodes/dataset/sqlite/queries.py (hunk 4 above)
Comment (WilliamHPNielsen) [reply qcodes/dataset/sqlite/queries.py]: Sure, we can use the pragma. It's a good idea to not just catch-all on `RuntimeError`.

A `NULL` value retrieved as a `None`. The case when `maybe_str_link is None` is handled a few lines down.

File: qcodes/dataset/measurements.py (hunk 3 above)
Comment (jenshnielsen) [reply qcodes/dataset/measurements.py]: I think its a good idea to normalize but on the other hand i have no idea of the relevant types so lets leave it for now

File: qcodes/dataset/sqlite/queries.py (hunk 4 above)
Comment (jenshnielsen) [reply qcodes/dataset/sqlite/queries.py]: Yes either A more specific exception catching or the pragma is fine with me

File: qcodes/dataset/linked_datasets/links.py (hunk 2 above)
Comment (WilliamHPNielsen) [reply qcodes/dataset/linked_datasets/links.py]: Yes, let's do that. Note that the verification is only of the format. Nonsensical values can still slip through.

Comment (jenshnielsen) [reply qcodes/dataset/linked_datasets/links.py]: Yes but that can be improved in a future PR

**Pull request merged**